### DISCUSSION TITLE
Pass Handles Array to Custom Node?

### DISCUSSION THREAD

**Discussion by tgram-3D** [xyflow/xyflow#4469] [Q&A] [answered]
I'm trying to figure out how to dynamically add and remove handles to a custom node efficiently. Currently the NodeProps type does not allow passing in a handles array, but the node.handles prop is still accessible from the actual nodes state. 
```

const updateNodesWithHandles = (nodes: CustomNode[]) => {
        return nodes.map(node => {
            node.handles = node.handles || [];
            const { x, y } = node.position;
            node.handles.push({
                id: "a",
                position: Position.Top,
                type: "target",
                x: x,
                y: y
            });
            node.handles.push({
                id: "a",
                position: Position.Bottom,
                type: "source",
                x: x,
                y: y
            });

            return node;
        });
    }
```
Long story short, if I do something like the above and do setNodes with the result, I can access node.handles from the nodes state, but I CANNOT access props.handles or something within the actual CustomNode component itself to render the handles. As a result I'm doing some hacky solution where I'm conditionally rendering handles in certain nodes based on a boolean flag, and that runs into a race condition with updateNodeInternals since it takes longer for the handles to render than it does to change the flag state, so basically I have to set a timeout on setEdges to eliminate the "no id" warning. 

I also noticed that the NodeHandles type is not exported from react-flow for whatever reason. 

It seems like there should be a way to do this. 

```
export interface CustomNodeData extends Record<string, unknown> {
    node: string;
    isLoading: boolean;
    args?: FinalNode;
}

export type CustomNode = Node<CustomNodeData>;

export const GraphNode = (props: NodeProps<CustomNode>) => {
```

I'm new to ReactFlow so forgive me if I'm missing something obvious. I know I could add a handles prop to CustomNodeData or something but I was hoping the built-in node.handles prop and NodeHandle type could be used instead somehow.

**Comment by bcakmakoglu** [comment, marked as answer]
Mh... actually not sure why the handles aren't passed as a prop, whether that's intentional or not, but if you want to use `node.handles` and still get them in your custom node component you can use the `useInternalNode` hook.

```ts
const internalNode = useInternalNode(props.id)

const handles = internalNode?.handles || []
```

**Reply by tgram-3D** [reply]
Thank you for the reply! The useInternalNode hook works for what I was trying to do. 

This took me way too long to figure out, but I was also able to eliminate the  `Couldn't create edge for source handle id: ` console warning just by clearing the previous edges before updating nodes/handles and calling updateNodeInternals().  

I was surprised handles wasn't available as a NodeProp, since it seems like a convenient way to conditionally render handles vs. defining it all in jsx in the custom node component, but this approach works just fine. Thanks again.
